### ISSUE TITLE
InMemoryDatabase raises System.InvalidOperationException when you save TPH Model

### ISSUE THREAD

**Issue by katz** [aspnet/EntityFramework#4285]
```C#
Unhandled Exception: System.InvalidOperationException: The property 'Discriminator' on entity type 'SubProduct' has a temporary value while attempting to change the entity's state to 'Unchanged'. Either set a permanent value explicitly or ensure that the database is configured to generate values for this property.
   at Microsoft.Data.Entity.ChangeTracking.Internal.InternalEntityEntry.SetEntityState(EntityState oldState, EntityState newState, Boolean acceptChanges)
   at Microsoft.Data.Entity.ChangeTracking.Internal.InternalEntityEntry.SetEntityState(EntityState entityState, Boolean acceptChanges)
   at Microsoft.Data.Entity.ChangeTracking.Internal.InternalEntityEntry.AcceptChanges()
   at Microsoft.Data.Entity.ChangeTracking.Internal.StateManager.AcceptAllChanges(IReadOnlyList`1 changedEntries)
   at Microsoft.Data.Entity.ChangeTracking.Internal.StateManager.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.Data.Entity.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.Data.Entity.DbContext.SaveChanges()
```

The below is a sample console application that raises the exception.
The sample app uses EF 7.0.0-rc1-final with .NET Framework 4.6.

```C#
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.Data.Entity;
using Microsoft.Data.Entity.Infrastructure;
using Microsoft.Data.Entity.Metadata;
using System.ComponentModel.DataAnnotations;

namespace EF7ConsoleApplication
{
    class Program
    {
        static void Main(string[] args)
        {
            var optionsBuilder = new DbContextOptionsBuilder<ApplicationDbContext>();
            // This is the magic line
            optionsBuilder.UseInMemoryDatabase();
            var db = new ApplicationDbContext(optionsBuilder.Options);

            var subProductList = new List<SubProduct>()
            {
                new SubProduct()
                {
                    someName = "test1",
                },
                new SubProduct()
                {
                    someName = "test2",
                },
            };

            db.AddRange(subProductList);
            // The SaveChanges() raises the exception.
            db.SaveChanges();
        }
    }

    public class BaseProduct
    {
        [Key]
        public Guid ProductId { get; set; }
    }

    public class SubProduct : BaseProduct
    {
        public string someName { get; set; }
    }
    public class SubProduct2 : BaseProduct
    {

    }

    public class ApplicationDbContext : DbContext
    {
        public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options) : base(options) { }
        
        public DbSet<SubProduct> SubProducts { get; set; }
        public DbSet<SubProduct2> SubProduct2 { get; set; }

        protected override void OnModelCreating(ModelBuilder builder)
        {
            base.OnModelCreating(builder);

            builder.Entity<BaseProduct>().ToTable("BaseProducts")
                .HasDiscriminator<string>("Discriminator")
                .HasValue<SubProduct>("SubProduct")
                .HasValue<SubProduct2>("SubProduct2");
        }
    }
}
```

**Comment by katz** [comment]
If you remove the below code, the exception isn't raised.

```C#
builder.Entity<BaseProduct>().ToTable("BaseProducts")
    .HasDiscriminator<string>("Discriminator")
    .HasValue<SubProduct>("SubProduct")
    .HasValue<SubProduct2>("SubProduct2");
```

**Comment by rowanmiller** [comment]
**Notes for Triage:** In triage we weren't sure whether this would be an issue specific to InMemory. When targeting a relational database the discriminator value is correctly set before SaveChanges but when targeting InMemory the discriminator property is set to a temporary value (string representation of a GUID - as is normal for temporary string values). The problem is that the `HasDiscriminator` and `HasValue` APIs are relational specific but the discriminator column is still being added to the model as a required property (enter long discussion about needing structural model changes when targeting different providers :smile:)

**Comment by rowanmiller** [comment]
We need to come up with a solution for this, possibly a way for HasDiscriminator to set a value generator to be used for the property regardless of what provider is being targeted.

**Comment by jnm2** [comment]
I'm up against the same issue. It's very disappointing that InMemory does not have a fundamental understanding of relational mappings. I get the point of requiring structural model changes when targeting different providers, but isn't the whole point of InMemory to test? How are relational models intended to be tested?

Although you could argue that TPH inheritance is not fundamentally a relational concept (TPH is not a storage concept) and move HasDiscriminator to EntityFramework.Core and make me even happier... 

TPT is the only thing that requires relational storage. TPC doesn't require relational storage either- I'm assuming InMemory uses TPC under the covers in this scenario.

**Comment by rowanmiller** [comment]
@jnm2 we agree... this is on our list to fix before v1 RTM

**Comment by jnm2** [comment]
You agree about `HasDiscriminator` being available for nonrelational? Fantastic!

**Comment by rowanmiller** [comment]
@jnm2 agreed that it should all work... we're not quite sure what the implementation will look like. The discriminator column isn't really needed for InMemory, so we may use it anyway... or just come up with some approach where it doesn't cause issues by being there.

**Comment by jnm2** [comment]
Okay. I know this is wider topic but I'm really hoping for TPH on non-relational providers other than InMemory as well.

**Comment by ajcvickers** [comment]
Fixed with #5090. Marking for re-triage to discuss if we want to do anything else to handle structural changes introduced only for use by some providers.
